### PR TITLE
Show warning in eager mode for empty containers

### LINKED ISSUE #54873: `torch.jit.isinstance` is overly admissive when comparing container types in eager mode
## 🐛 Bug

`torch.jit.isinstance` will falsely return `True` in eager when comparing any two container types that are the same, even if those types don't have the same subscript. My guess is that we silently transform `List[T]` into `list` for the eager mode check, which is not correct. The example below demonstrates that `torch.jit.isinstance` recognizes `List[T1]` as an instance of `List[T2]` since both types are `List`s, even if `T1` and `T2` are invariant.

This is not a problem in script mode.

## Example

```python
import torch
from typing import List

def fn():
    x: List[torch.Tensor] = []
    if torch.jit.isinstance(x, List[int]):
        print("Wrong!")
    else:
        print("Right")
    return x

scripted = torch.jit.script(fn)
scripted()                 # prints "Right"
fn()                       # prints "Wrong!"
```

cc @gmagogsfm

### PR DESCRIPTION
**Pull request by karolsputo**
Fixes #54873

### CODE REVIEW

**Review by ansley — CHANGES_REQUESTED** [review]
Looks great! I left a small note about the wording of your warning, but it's otherwise ready to go. Thanks for contributing!

File: torch/_jit_internal.py [hunk 1]
```diff
@@ -1125,12 +1125,18 @@ def check_args_exist(target_type) -> None:
         raise_error_container_parameter_missing("Optional")
 
 
+def check_empty_containers(obj) -> None:
+    if not obj:
+        warnings.warn("Type annotations are lost in eager mode. False posivites will be shown for empty containers.")
```
[the remark this answers is not in the document]
Comment (karolsputo) [reply torch/_jit_internal.py]: I see how that's accurate, thanks. I've changed it slightly, hope you don't mind.

**Review by karolsputo — COMMENTED** [review]
Fixed the warning text.

File: torch/_jit_internal.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ansley) [reply torch/_jit_internal.py]: Perfect! I like your phrasing a lot better...mine was too wordy :)

**Review by ansley — APPROVED** [review]
Great job, Karol! Thanks for working on these OSS issues. I'll try to keep my eye out for some other good ones you could do, if you'd like. Do you have an area of focus or special interest?

**Comment by karolsputo** [comment]
I'm currently trying to get a feel for the whole codebase so no specific areas. I'm taking a look at some DataParallel issues right now for a change. I'm interested in AI frameworks and training hardware so I'd like to progressively get closer to the core and hardware stuff. Taking small steps :)
